File: frontend/src/lib/mutations.ts
import type { ExplorationMutation, ExplorationSession } from '../types/exploration'

export function applyExplorationMutations(
  session: ExplorationSession,
  mutations: ExplorationMutation[],
): ExplorationSession {
  const next: ExplorationSession = {
    ...session,
    nodes: [...session.nodes],
    edges: [...session.edges],
    runs: [...session.runs],
    favorites: [...session.favorites],
  }

  for (const mutation of mutations) {
    switch (mutation.kind) {
      case 'node_added':
        if (mutation.node && !next.nodes.some((node) => node.id === mutation.node!.id)) {
          next.nodes.push(mutation.node)
        }
        break
      case 'edge_added':
        if (mutation.edge && !next.edges.some((edge) => edge.id === mutation.edge!.id)) {
          next.edges.push(mutation.edge)
        }
        break
      case 'run_added':
        if (mutation.run && !next.runs.some((run) => run.id === mutation.run!.id)) {
          next.runs.push(mutation.run)
        }
        break
      case 'favorites_updated':
        next.favorites = mutation.favorites ? [...mutation.favorites] : []
        break
      case 'active_opportunity_set':
        if (mutation.active_opportunity_id) {
          next.activeOpportunityId = mutation.active_opportunity_id
        }
        break
      case 'strategy_updated':
        if (mutation.strategy) {
          next.strategy = mutation.strategy
        }
        break
      default:
        break
    }
  }

  return next
}

File: backend/domain/exploration/mutations.go
package exploration

import (
	"fmt"
	"slices"
	"time"
)

func mutationID(workspaceID string) string {
	return fmt.Sprintf("mutation-%s-%d", workspaceID, time.Now().UnixNano())
}

func diffMutations(prev ExplorationSession, next ExplorationSession, source string) []MutationEvent {
	out := make([]MutationEvent, 0)

	if prev.ActiveOpportunityID != next.ActiveOpportunityID {
		out = append(out, MutationEvent{
			ID:                  mutationID(next.ID),
			WorkspaceID:         next.ID,
			Kind:                string(MutationKindActiveOpportunitySet),
			Source:              source,
			ActiveOpportunityID: next.ActiveOpportunityID,
			CreatedAt:           time.Now().UnixMilli(),
		})
	}

	prevNodeIDs := map[string]struct{}{}
	for _, node := range prev.Nodes {
		prevNodeIDs[node.ID] = struct{}{}
	}
	for _, node := range next.Nodes {
		if _, ok := prevNodeIDs[node.ID]; ok {
			continue
		}
		copyNode := node
		out = append(out, MutationEvent{
			ID:          mutationID(next.ID),
			WorkspaceID: next.ID,
			Kind:        string(MutationKindNodeAdded),
			Source:      source,
			Node:        &copyNode,
			CreatedAt:   time.Now().UnixMilli(),
		})
	}

	prevEdgeIDs := map[string]struct{}{}
	for _, edge := range prev.Edges {
		prevEdgeIDs[edge.ID] = struct{}{}
	}
	for _, edge := range next.Edges {
		if _, ok := prevEdgeIDs[edge.ID]; ok {
			continue
		}
		copyEdge := edge
		out = append(out, MutationEvent{
			ID:          mutationID(next.ID),
			WorkspaceID: next.ID,
			Kind:        string(MutationKindEdgeAdded),
			Source:      source,
			Edge:        &copyEdge,
			CreatedAt:   time.Now().UnixMilli(),
		})
	}

	if len(next.Runs) > len(prev.Runs) {
		for _, run := range next.Runs[len(prev.Runs):] {
			copyRun := run
			out = append(out, MutationEvent{
				ID:          mutationID(next.ID),
				WorkspaceID: next.ID,
				Kind:        "run_added",
				Source:      source,
				Run:         &copyRun,
				CreatedAt:   time.Now().UnixMilli(),
			})
		}
	}

	if !slices.Equal(prev.Favorites, next.Favorites) {
		out = append(out, MutationEvent{
			ID:          mutationID(next.ID),
			WorkspaceID: next.ID,
			Kind:        string(MutationKindFavoritesUpdated),
			Source:      source,
			Favorites:   append([]string{}, next.Favorites...),
			CreatedAt:   time.Now().UnixMilli(),
		})
	}

	if prev.Strategy != next.Strategy {
		strategy := next.Strategy
		out = append(out, MutationEvent{
			ID:          mutationID(next.ID),
			WorkspaceID: next.ID,
			Kind:        string(MutationKindStrategyUpdated),
			Source:      source,
			Strategy:    &strategy,
			CreatedAt:   time.Now().UnixMilli(),
		})
	}

	return out
}
